File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        
        HashMap<Character, Integer> letters = new HashMap<>();
        
        for(int i=0; i < magazine.length(); i++) {
            char c = magazine.charAt(i);
            
            int count = letters.getOrDefault(c, 0);
            
            letters.put(c, count+1);
        }
        
        for(int i=0; i < ransomNote.length(); i++) {
            char c = ransomNote.charAt(i);
            
            int count = letters.getOrDefault(c, 0);
            
            if(count == 0) {
                return false;
            }
            
            letters.put(c, count-1);
        }
        return true;
        
    }
}

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        
        if len(ransomNote) > len(magazine):
            return False
        
        mySet = set(ransomNote)

        for c in mySet:
            if ransomNote.count(c) > magazine.count(c):
                return False

        return True

        # # slow solution
        # hashMap = {}

        # for c in magazine:
        #     if c in hashMap:
        #         hashMap[c] += 1
        #     else:
        #         hashMap[c] = 1
        
        # for c in ransomNote:
        #     if c in hashMap and hashMap[c] > 0:
        #         hashMap[c] -= 1
        #     else:
        #         return False
        
        # return True